### DISCUSSION TITLE
usePrepareContractWrite automaticly get called before click!

### DISCUSSION THREAD

**Discussion by arash-hk** [wevm/wagmi#2037] [Q&A] [answered]
### hi everyone, im using web3modal and wagmi and i created a few simple buttons with nextjs and im trying to interact with my smart contract, but when i use usePrepareContractWrite before i pass my config to useContractWrite hook and call it with a button, it automaticly get called in my nextjs app when page refreshed!

can anyone help me please?

here is my code:

1. my app page that i did my config for wagmi:

![Capture](https://user-images.githubusercontent.com/111232556/226342018-2195841f-8b1e-4b77-873f-38d73872fdd2.PNG)

2.and my index file witch has different parts

A: IMPORTS

![Capture](https://user-images.githubusercontent.com/111232556/226343380-16fe46ff-12c1-4a51-adce-f80f9d896504.PNG)

B: web3modal v2 wallet theme

![Capture2](https://user-images.githubusercontent.com/111232556/226343574-49c2b045-c8ac-4720-873a-290ac28fe2db.PNG)

C : usePrepareContractWrite with an extended example

![Capture3](https://user-images.githubusercontent.com/111232556/226343723-f5843065-78f5-43cc-8710-0dc855f84a3e.PNG)

D: useContracWrite and jsx part witch is useless here!

![Capture4](https://user-images.githubusercontent.com/111232556/226344034-9c0931af-7162-4693-bddb-fba68f42c2d8.PNG)

E: and here it is! when i run "yarn dev" and start my server, page refreshes and in console tab you will see this messages! **the errors are common we are handling them in frontend part ! the main question is why usePrepareContractWrite automactily do this? i want to have it with onClick !?**

![Capture6](https://user-images.githubusercontent.com/111232556/226344607-db318070-d888-4914-99d6-1b3ff5a35d35.PNG)

and here is my json package, thanks for your attention!

![pck](https://user-images.githubusercontent.com/111232556/226344953-246efb5b-43a2-45ee-b90c-de81db193de8.PNG)

**Comment by SaadNiaziDev** [comment, marked as answer]
if u want to run this onClick, use enabled: false in prepareWriteHook and onClick do 
``
const mint = async() =>{
await refetch();
write?.();
};
``
refetch is function in usePrepareHook

**Reply by arash-hk** [reply]
thank you so much dear saad!

i sticking in this problem about 3 days!

i share the proper code for guys who have the same problem!

here is the code:

![Capture](https://user-images.githubusercontent.com/111232556/227166677-42985a9f-af7b-4cf7-9bff-3d06dc110f02.PNG)

**make sure to import `refetch` in `usePrepareContractWrite`**

**Reply by 0xmuneeb** [reply]
You guys are love💖 I was stuck in this problem and was not able to find any solution for few days. Thank you so much

**Reply by arash-hk** [reply]
dear muneeb there is point you have to now, after i solved this problem i faced with an another issue is IOS devices, and i started another discusion with friends to figure out what is going on! 
so i recommand to study this link also: https://github.com/wagmi-dev/wagmi/discussions/2115#discussioncomment-5546905

best regards!

**Reply by 0xmuneeb** [reply]
Thank you so much for mentioning this as well! I read that thread as well, so that means I should not be using `refetch`  and `enabled: false` but there is also a problem with my project, I don't want to disable the mint button and display error automatically if the user does not have any funds!! what should I do then? Anyways I'll figure something out!

Again thanks a lot 
best regards:
_Muneeb Ur Rehman_

**Reply by arash-hk** [reply]
hello again dear muneeb, 
you can use some frontend stuff to return some client friendly responses to client when button is disabled! you can handle and send to client with variable ways, you can use toastify, and syntaxy you can say if disabled was true --- send client "you dont have enough amount of ETH" (something like this !)
https://www.npmjs.com/package/react-toastify
you can read about this npm library link above, if you had more queestion, i am here to answere

**Reply by 0xmuneeb** [reply]
okay I understand! Thank you so much for your help and time.
